include/deal.II/non_matching/quadrature_generator.h

```diff
@@ -0,0 +1,236 @@
+         * same way as for a cell.
+         */
+        Hypercube<dim>
+        child(const unsigned int index) const;
+
+      private:
+        /**
+         * Coordinate in the center of the hypercube.
+         */
+        const Point<dim> center_coordinate;
+
+        /**
+         * Side length of the hypercube.
+         */
+        const double length_of_side;
+      };
+
+
+      /**
+       * This function defines a convention for how coordinates in dim
+       * dimensions should translate to the coordinates in dim + 1 dimensions,
+       * when one of the coordinates in dim + 1 dimensions is locked to a given
+       * value.
+       *
+       * The convention is the following: Starting from the locked coordinate we
+       * store the lower dimensional coordinates consecutively and wrapping
+       * around when going over the dimension. This relationship can be
+       * described by the following tables:
+       *
+       *                        2D
+       * ------------------------------------------------
+       * | locked in 2D | 1D coordinate | 2D coordinate |
+       * |--------------------------------------------- |
+       * |     x0       |      (a)      |   (x0,  a)    |
+       * |     x1       |      (a)      |   (a , x1)    |
+       * ------------------------------------------------
+       *
+       *                       3D
+       * --------------------------------------------------
+       * | locked in 3D | 2D coordinates | 3D coordinates |
+       * |----------------------------------------------- |
+       * |     x0       |    (a, b)      | (x0,  a,  b)   |
+       * |     x1       |    (a, b)      | ( b, x1,  a)   |
+       * |     x2       |    (a, b)      | ( a,  b, x2)   |
+       * --------------------------------------------------
+       *
+       * Given a locked coordinate, this function maps a coordinate index in dim
+       * dimension to a coordinate index in dim + 1 dimensions.
+       *
+       * @param locked_coordinate should be in the range [0, dim+1).
+       * @param coordiante_in_dim should be in the range [0, dim).
+       * @return A coordinate index in the range [0, dim+1)
+       */
+      template <int dim>
+      inline unsigned int
+      coordinate_to_one_dim_higher(const unsigned int locked_coordinate,
+                                   const unsigned int coordiante_in_dim)
+      {
+        AssertIndexRange(locked_coordinate, dim + 1);
+        AssertIndexRange(coordiante_in_dim, dim);
+        return (locked_coordinate + coordiante_in_dim + 1) % (dim + 1);
+      }
+
+
+
+      /* -------------- declaration of explicit specializations ------------- */
+
+      template <>
+      typename Hypercube<0>::CrossSection
+      Hypercube<0>::cross_section(const unsigned int) const;
+
+      template <>
+      Point<0>
+      Hypercube<0>::vertex(const unsigned int) const;
+
+      template <>
+      Hypercube<0>
+      Hypercube<0>::child(const unsigned int) const;
```

Comment:
I don't think these specializations are in the patch any more. If they aren't then would you please remove them?

Reply:
I think I need these to make it compile. It's analogous as the declarations for `Quadrature<0>` at the end of quadrature.h. Now I have added
```
template class Hypercube<0>;
```
in the .cc-file.

`Hypercube<dim>::cross_section(..)` creates an object of type `Hypercube<dim-1>`  
so I need to specialise this function for `dim==0` to avoid that the compiler creates 
`Hypercube<-1>`. 

Both `Hypercube<dim>::vertex(..)` and `Hypercube<dim>::child(..)`  
uses `GeometryInfo<dim>::unit_cell_vertex` which doesn't exist for `dim==0`.
If I remove it I get
`error: ‘unit_cell_vertex’ is not a member of ‘dealii::GeometryInfo<0>’` 

I guess the alternative is to not explicitely template `Hypercube<0>`. I'm not sure what the best option is. :thinking:

Reply:
But yes, there is no implementation for these functions in the .cc-file.

Reply:
I think the lack of an implementation was what confused me. I'll take a closer look later today.

Reply:
> But yes, there is no implementation for these functions in the .cc-file.

Only providing a forward declaration without any implementation is asking for problems... I would at least expect some undefined references if you really need it.

I would just implement them with an `ExcNotImplemented` assert.

Reply:
> I would just implement them with an `ExcNotImplemented` assert.

I can do that for two of the functions, but `Hypercube<0>::cross_section` should return a
`Hypercube<-1>`. If I add an implementation the compiler instantiates `Hypercube<-1>` which fails to compile.


> Only providing a forward declaration without any implementation is asking for problems...

Perhaps, but this workaround was based on how it's done in the Quadrature class. This constructor is declared in quadrature.h but does not have an implementation:
```
template <>
Quadrature<0>::Quadrature(const Quadrature<-1> &, const Quadrature<1> &);
```

Reply:
I think this strategy is fine - it works for Quadrature and it works here too.